### PR TITLE
Factor registry

### PR DESCRIPTION
**Pull request by vcaputo**

### CODE REVIEW

File: exec/stages/stages.go [hunk 1]
```diff
@@ -31,29 +30,20 @@ type StageCreator interface {
...
 func Get(name string) StageCreator {
-	if stage, ok := stages[name]; ok {
-		return stage
+	s := stages.Get(name)
+	if s == nil {
+		return nil
 	}
-
-	return nil
+	return s.(StageCreator)
```
Comment (crawford) [inline exec/stages/stages.go]: Ugh, this is hideous without generics/templates.

Comment (vcaputo) [inline exec/stages/stages.go]: Less hideous than copy and paste implementations of the same thing IMO.

We still have type-safety at the stages and provider Register/Get boundaries, and the Registry instances are private to the packages, so I don't really see the problem.

Comment (crawford) [inline exec/stages/stages.go]: Yeah, this should be fine. When I looked at this last (on my phone), I couldn't see the full context.

**Comment by jonboulle** [comment]
useless $0.02:
I'm struggling to justify the general registry here and associated typecasting. Is the expectation that this evolves to support multithreaded access? My initial reaction is that even the original Get/Names approach is overcomplicated; a simple `func Names(map[string]interface{}) []string` would suffice generically for the latter, and `Get` doesn't appear to be providing anything over just accessing the map.

**Comment by vcaputo** [comment]
I don't feel strongly about the Register/Get/Names approach, but I do prefer it to packages operating directly on public maps.  It may be slightly over-engineered but nothing egregious.

If we keep the current approach, I do think the generic registry proposed here improves upon its implementation.

@crawford what's the rationale for the existing stages and providers apis?

**Comment by crawford** [comment]
I wanted to allow packages to register themselves in the "database" but not modify others. Exposing the entire map allows people too much access.

**Comment by jonboulle** [comment]
People? What people? ;-)

**Comment by jonboulle** [comment]
img

**Comment by vcaputo** [comment]
Funny, but can I get an lgtm? :)

File: Makefile [hunk 2]
```diff
@@ -61,4 +62,4 @@ fix:
 	$(Q)go tool fix -diff $(PACKAGES)
 test: | gopath/src/github.com/coreos/ignition
 	@echo " TEST  $(PACKAGES)"
-	$(Q)GOPATH=$$(pwd)/gopath go test $(ABS_PACKAGES)
+	$(Q)GOPATH=$$(pwd)/gopath go test -cover $(ABS_PACKAGES)
```
Comment (jonboulle) [inline Makefile]: does this actually work with multiple packages now?

Comment (jonboulle) [inline Makefile]: ohh, nvm, I'm thinking of `go tool cover`

File: registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+
+package registry
+
+import (
+	"fmt"
+	"sort"
+)
+
+// Types implementing the Registrant interface may be registered
```
Comment (jonboulle) [inline registry/registry.go]: // Registrant ...

File: registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+type Registrant interface {
+	Name() string
+}
+
+type Registry struct {
+	name        string
```
Comment (jonboulle) [inline registry/registry.go]: Why does a Registry need to be named?

Comment (jonboulle) [inline registry/registry.go]: or rather, need to introspect its name - debugging?

Comment (jonboulle) [inline registry/registry.go]: never mind, how about I read the code

File: registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+// Register registers a new registrant to a registry
+func (r *Registry) Register(registrant Registrant) {
+	if _, ok := r.registrants[registrant.Name()]; ok {
+		panic(fmt.Sprintf("%s: registrant %q already registered", r.name, registrant.Name()))
+	}
+	r.registrants[registrant.Name()] = registrant
+}
+
+// Get gets an named registrant from a registry
```
Comment (jonboulle) [inline registry/registry.go]: s/an/a/

**Comment by jonboulle** [comment]
looks _okay_ to me ;-)

File: registry/registry.go (hunk 3 above)
Comment (vcaputo) [inline registry/registry.go]: As in s/Types implementing the //?

Comment (jonboulle) [inline registry/registry.go]: yep, the godoc "Name foos the bar" or "Name should be implemented to woof the miaow" or whatever

**Comment by vcaputo** [comment]
I'll take an lotm.

**Pull request merged**